### PR TITLE
perf: fix space leaks by downgrading fuzzyset to v0.2.4

### LINKED ISSUE #3329: Stack Overflow error in dev enviroment when table and embed resource do not exist
### Environment

* PostgreSQL version: 16.1
* PostgREST version: main branch
* Operating system: NixOS 23.11

### Description of issue

When requesting a table that does not exist with any embedded resource, it shows a `stack overflow` error in the logs and fails the request. To reproduce:

```console
$ nix-shell --pure --run "postgrest-with-postgresql-16 -f ./test/spec/fixtures/load.sql postgrest-run"
```
```console
$ curl "localhost:3000/x?select=y(*)"
curl: (52) Empty reply from server
```
Expected:
```
$ curl "localhost:3000/x?select=y(*)"
{
  "code":"PGRST200",
  "details":"Searched for a foreign key relationship between 'x' and 'y' in the schema 'public', but no matches were found.",
  "hint":null,
  "message":"Could not find a relationship between 'x' and 'y' in the schema cache"
}
```

This only happens in the Nix development environment  for now (not for spec tests), since it enables the leak detection with `-K1K`:

https://github.com/PostgREST/postgrest/blob/ec7ab271d69e8ad907328b34c7d1a4aed1c60af9/postgrest.cabal#L185

It appears that the `fuzzyset` package, used to build the "hint" of the error, is the culprit. The evidence is that it is marked as "broken" in the latest update of nixpkgs/GHC in commit 6682baeb (previous commits do not throw the error even with `-K1K`):

https://github.com/PostgREST/postgrest/blob/ec7ab271d69e8ad907328b34c7d1a4aed1c60af9/nix/overlays/haskell-packages.nix#L48-L49

### PR DESCRIPTION
**Pull request by laurenceisla**
Closes #3329.

### CODE REVIEW

**Comment by steve-chavez** [comment]
@laurenceisla How about a test for this? We now have the `test/io/test_big_schema.py`. You should be able to reproduce the `stack overflow` there, I think.

File: test/io/test_big_schema.py [hunk 1]
```diff
@@ -70,3 +70,21 @@ def test_openapi_in_big_schema(defaultenv):
     with run(env=env) as postgrest:
         response = postgrest.session.get("/")
         assert response.status_code == 200
+
+
+# See: https://github.com/PostgREST/postgrest/issues/3329
+def test_should_not_fail_with_stack_overflow(defaultenv):
```
Comment (laurenceisla) [inline test/io/test_big_schema.py:76]: This test fails on the `main` branch with "stack overflow".

**Comment by steve-chavez** [comment]
> Don't know the details, but it's similar in that it takes more space in memory than it's necessary, but it's freed later.

Just to clarify. It's not a memory leak, is just that we limit the stack size during development to `1K` with:

https://github.com/PostgREST/postgrest/blob/0d4441852e1c2383227f03ee189e563c611b0966/postgrest.cabal#L185

Which forces a `stack overflow` error.

More details at https://neilmitchell.blogspot.com/2015/09/detecting-space-leaks.html?m=1

---

Don't know `fuzzyset` details and why it consumes noticeable memory now. It'd be better to open an issue upstream to see if it can be fixed.

**Comment by steve-chavez** [comment]
> What is a "space leak"? 

Actually the term was correctly used:

> A space leak occurs when a computer program uses more memory than necessary. In contrast to memory leaks, where the leaked memory is never released, the memory consumed by a space leak is released, but later than expected.

References:

- https://queue.acm.org/detail.cfm?id=2538488
- https://stackoverflow.com/questions/46007746/whats-a-space-leak
- https://www.reddit.com/r/haskell/comments/pvosen/how_can_haskell_programmers_tolerate_space_leaks/
- https://chshersh.com/space-leak
- https://softwareengineering.stackexchange.com/questions/306226/does-it-make-sense-to-use-the-term-space-leak-with-regard-to-java

**Comment by laurenceisla** [comment]
> But maybe we can / should downgrade temporarily while still reporting the problem upstream.

Agree, opened an issue there: https://github.com/laserpants/fuzzyset-haskell/issues/9

**Comment by taimoorzaeem** [comment]
@laurenceisla With this downgrade, `postgrest-test-doctests` fails with the following:
```
src/PostgREST/Error.hs:294:26: error:
    Not in scope: ‘Fuzzy.fromList’
    NB: the module ‘Data.FuzzySet’ does not export ‘fromList’.

src/PostgREST/Error.hs:295:26: error:
    Not in scope: ‘Fuzzy.fromList’
    NB: the module ‘Data.FuzzySet’ does not export ‘fromList’.

src/PostgREST/Error.hs:296:21: error:
    Not in scope: ‘Fuzzy.getOne’
    NB: the module ‘Data.FuzzySet’ does not export ‘getOne’.

src/PostgREST/Error.hs:298:43: error:
    Not in scope: ‘Fuzzy.get’
    NB: the module ‘Data.FuzzySet’ does not export ‘get’.
.
.
.
```

**Comment by taimoorzaeem** [comment]
> Strange. Why did this not fail in CI. I think this was the latest run here: https://github.com/PostgREST/postgrest/actions/runs/8349585526/job/22854079725. Doctests seem fine here.

On more thought, I think it is some local issue on my side. My bad.

**Comment by laurenceisla** [comment]
@steve-chavez Strange, I'm not getting those errors:

```console
[nix-shell:~/Projects/postgrest]$ postgrest-test-doctests
Warning: Both /home/laurence/.cabal and /home/laurence/.config/cabal/config
exist - ignoring the former.
It is advisable to remove one of them. In that case, we will use the remaining
one by default (unless '$CABAL_DIR' is explicitly set).

src/PostgREST/Version.hs:12:1: error:
    Could not find module ‘Paths_postgrest’
    Perhaps you meant Paths_doctest
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
Examples: 153  Tried: 153  Errors: 0  Failures: 0
```

Maybe your nix-shell is caching something? `postgrest-clean` could help.

**Comment by steve-chavez** [comment]
Ah, right. Just reentering `nix-shell` made those errors go.

**Pull request merged**